### PR TITLE
feat: Add `InputConversion` & `OutputConversion` for nn interface

### LINKED ISSUE #621: feat: add Input Layer for nn Interface
### Is your feature request related to a problem?

We are right now working on translating the safe-ds data into pytorch tensors. The translation is defined in the dataclasses them self, but the definiton how, should be defined on the input layer interface.

### Desired solution

Create Input Layer Interface, with Type parameter, so we have static type analysis.

### PR DESCRIPTION
**Pull request by Gerhardsa0**
Closes #621 

### Summary of Changes

added input and output conversion Interface
also added this functionality for Table

### CODE REVIEW

File: src/safeds/ml/nn/_input_conversion.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+from abc import ABC, abstractmethod
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_input_conversion.py:1]: Add `from __future__ import annotations`

File: src/safeds/ml/nn/_input_conversion.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+from typing import Generic, TypeVar
+
+from torch.utils.data import DataLoader
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_input_conversion.py:4]: Move into a `TYPE_CHECKING` block.

File: src/safeds/ml/nn/_input_conversion_table.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+from torch.utils.data import DataLoader
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_input_conversion_table.py:1]: Future/annotations + TYPE_CHECKING.

File: src/safeds/ml/nn/_output_conversion.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
+from abc import ABC, abstractmethod
+from typing import Generic, TypeVar
+
+from torch import Tensor
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_output_conversion.py:4]: Annotations, TYPE_CHECKING.

File: src/safeds/ml/nn/_output_conversion_table.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+from torch import Tensor
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_output_conversion_table.py:1]: Annotations + TYPE_CHECKING.

File: src/safeds/ml/nn/_forward_layer.py [hunk 6]
```diff
@@ -1,7 +1,7 @@
 from torch import Tensor, nn
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_forward_layer.py:1]: Should not be imported globally. This single line of code takes 1.2s to run.

File: src/safeds/ml/nn/_input_conversion.py (hunk 2 above)
Comment (Marsmaennchen221) [reply src/safeds/ml/nn/_input_conversion.py:4]: All of these or just the `DataLoader`?

Comment (lars-reimann) [reply src/safeds/ml/nn/_input_conversion.py]: Just the `DataLoader`. It's fine to load globally

1. our stuff (`safeds...`),
2. anything built into Python.

If you add the `from __future__ import annotations`, `ruff` should also suggest moving this import into a `TYPE_CHECKING` block.

**Review by lars-reimann — CHANGES_REQUESTED** [review]

**Comment by sibre28** [comment]
We probably should add explicit test cases for the methods of InputConversionTable and OutputConversionTable. Currently these are only being covered by the model tests.

**Review by Marsmaennchen221 — APPROVED** [review]

**Comment by Gerhardsa0** [comment]
@lars-reimann your review is needed

**Pull request merged**